jax/_src/nn/initializers.py

```diff
@@ -155,24 +189,33 @@ def _complex_uniform(key, shape, dtype):
   theta = 2 * jnp.pi * random.uniform(key_theta, shape, real_dtype).astype(dtype)
   return r * jnp.exp(1j * theta)
 
-def _complex_truncated_normal(key, upper, shape, dtype):
+def _complex_truncated_normal(key: KeyArray, upper: Array,
+                              shape: core.NamedShape,
+                              dtype: DTypeLikeInexact) -> Array:
   """
   Sample random values from a centered normal distribution on the complex plane,
-  whose modulus is truncated to `upper`, and the variance before the truncation is one.
+  whose modulus is truncated to `upper`, and the variance before the truncation
+  is one.
   """
   key_r, key_theta = random.split(key)
   real_dtype = np.array(0, dtype).real.dtype
   dtype = dtypes._to_complex_dtype(real_dtype)
-  t = (1 - jnp.exp(jnp.array(-(upper ** 2), dtype))) * random.uniform(key_r, shape, real_dtype).astype(dtype)
+  t = ((1 - jnp.exp(jnp.array(-(upper ** 2), dtype)))
+       * random.uniform(key_r, shape, real_dtype)).astype(dtype)
```

Comment:
This seems to be the change that caused the issue – the new version is no longer compatible with strict dtype promotion.

Reply:
Very kind of you to track this down for me!  I wasn't careful enough when I did the rebase.  It is hopefully fixed now.